Library/brew.sh

```diff
@@ -140,12 +173,47 @@ case "$HOMEBREW_COMMAND" in
   --config)    HOMEBREW_COMMAND="config";;
 esac
 
-if [[ -f "$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh" ]] ; then
+if [[ -f "$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh" ]]
+then
   HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh"
-elif [[ -n "$HOMEBREW_DEVELOPER" && -f "$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh" ]] ; then
+elif [[ -n "$HOMEBREW_DEVELOPER" && -f "$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh" ]]
+then
   HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh"
 fi
 
+if [[ "$HOMEBREW_COMMAND" = "help" || "$1" = "-h" || "$1" = "--help" ]]
+then
+
+  # Display sub command help for `brew help foo` and `brew foo --help`
+  if [[ "$HOMEBREW_ARG_COUNT" -gt 0 ]]
+  then
+    if [[ "$HOMEBREW_COMMAND" = "help" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$1"
+    else
+      HOMEBREW_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND"
+    fi
+
+    if [[ -f "$HOMEBREW_COMMAND.sh" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_COMMAND.sh"
+    elif [[ -f "$HOMEBREW_COMMAND.rb" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_COMMAND.rb"
+    else
+      echo "Unknown command \`$@\`"
+      echo
+      echo "$HOMEBREW_HELP"
+      exit 1
+    fi
+    grep "#:" "$HOMEBREW_COMMAND" | cut -c3- | tr -d \` | sed "s/^\* //"
+    exit 0
+  else
+    echo "$HOMEBREW_HELP"
+    exit 0
+  fi
+fi
```

Comment:
> For bash command, I'm thinking to redirect help flag to `brew help <command>`.

That's totally fine and what I'm suggesting, too. But we still have to put the help text for a command, that is implemented in Bash, somewhere. And I think it would be preferable if it could live in the same file that also implements the command. The language-agnostic solution to do this is to put the documentation into a comment (as all languages relevant to us use/accept `#` as a comment character).